djangosaml2/backends.py

```diff
@@ -110,50 +104,21 @@ def authenticate(self, session_info=None, attribute_mapping=None,
         if not self.is_authorized(attributes, attribute_mapping):
             return None
 
-        user = None
-
         main_attribute = self.clean_user_main_attribute(saml_user)
 
-        user_query_args = {
-                django_user_main_attribute+django_user_main_attribute_lookup:
-                main_attribute}
-        user_create_defaults = {django_user_main_attribute: main_attribute}
-
         # Note that this could be accomplished in one try-except clause, but
         # instead we use get_or_create when creating unknown users since it has
         # built-in safeguards for multiple threads.
-        User = get_saml_user_model()
-        if create_unknown_user:
-            logger.debug('Check if the user "%s" exists or create otherwise',
-                         main_attribute)
-            try:
-                user, created = User.objects.get_or_create(
-                        defaults=user_create_defaults, **user_query_args)
-            except MultipleObjectsReturned:
-                logger.error("There are more than one user with %s = %s",
-                             django_user_main_attribute, main_attribute)
-                return None
-
-            if created:
-                logger.debug('New user created')
-                user = self.configure_user(user, attributes, attribute_mapping)
-            else:
-                logger.debug('User updated')
-                user = self.update_user(user, attributes, attribute_mapping)
-        else:
-            logger.debug('Retrieving existing user "%s"', main_attribute)
-            try:
-                user = User.objects.get(**user_query_args)
-                user = self.update_user(user, attributes, attribute_mapping)
-            except User.DoesNotExist:
-                logger.error('The user "%s" does not exist', main_attribute)
-                return None
-            except MultipleObjectsReturned:
-                logger.error("There are more than one user with %s = %s",
-                             django_user_main_attribute, main_attribute)
-                return None
+        return self.get_saml2_user(
+            create_unknown_user, main_attribute, attributes, attribute_mapping)
 
-        return user
+    def get_attribute_value(self, django_field, attributes, attribute_mapping):
+        saml_user = None
+        logger.debug('attribute_mapping: %s', attribute_mapping)
+        for saml_attr, django_fields in attribute_mapping.items():
+            if (django_field in django_fields and saml_attr in attributes):
```

Comment:
Thanks for refactoring this spaghetti already. And while you're at it, could you please also remove those useless parenthesis? Hurts my eyes...